lib/mongo/client.rb

```diff
@@ -855,14 +857,42 @@ def validate_new_options!(opts = Options::Redacted.new)
     end
 
     # Validates all options after they are set on the client.
-    # This method is intended to catch combinations of options which do are
+    # This method is intended to catch combinations of options which are
     # not allowed.
     def validate_options!
       if options[:write] && options[:write_concern] && options[:write] != options[:write_concern]
         raise ArgumentError, "If :write and :write_concern are both given, they must be identical: #{options.inspect}"
       end
     end
 
+    # Validates all authentication-related options after they are set on the client
+    # This method is intended to catch combinations of options which are not allowed
+    def validate_authentication_options!
+      if options[:user].nil? && [:mongodb_cr, :plain, :scram, :scram256].include?(options[:auth_mech])
+        raise Mongo::Auth::InvalidConfiguration.new("user is required for mechanism #{options[:auth_mech]}")
+      end
+
+      if options[:auth_mech] == :mongodb_x509
+        if ![:external, nil].include?(options[:auth_source])
+          raise Mongo::Auth::InvalidConfiguration.new("#{options[:auth_source]} is an invalid auth source for mechanism mongodb_x509; valid options are $external and nil")
+        elsif !options[:password].nil?
+          raise Mongo::Auth::InvalidConfiguration.new('password is not supported for mongodb_x509 auth mechanism')
+        end
+      end
+
+      if options[:auth_mech].nil?
+        if !options[:user].nil? && !options[:user].length
```

Comment:
`!options[:user].length` is always false, suggest  `options[:user].empty?` instead.